### PR TITLE
Adds optional serde support to datafusion-proto

### LINKED ISSUE #2889: JSON version of `display_indent()`
**Is your feature request related to a problem or challenge? Please describe what you are trying to do.**

A significant portion of my time on #2885 was spent manually formatting subqueries like:

```
Projection: #SUM(lineitem.l_extendedprice) / Float64(7) AS avg_yearly
  Aggregate: groupBy=[[]], aggr=[[SUM(#lineitem.l_extendedprice)]]
    Filter: #part.p_brand = Utf8("Brand#23") AND #part.p_container = Utf8("MED BOX") AND #lineitem.l_quantity < (Subquery: Projection: Float64(0.2) * #AVG(lineitem.l_quantity)
  Aggregate: groupBy=[[]], aggr=[[AVG(#lineitem.l_quantity)]]
    Filter: #lineitem.l_partkey = #part.p_partkey
      TableScan: lineitem)
      Inner Join: #lineitem.l_partkey = #part.p_partkey
        TableScan: lineitem
        TableScan: part
```

into

```
Projection: #SUM(lineitem.l_extendedprice) / Float64(7) AS avg_yearly
  Aggregate: groupBy=[[]], aggr=[[SUM(#lineitem.l_extendedprice)]]
    Filter: #part.p_brand = Utf8("Brand#23") AND #part.p_container = Utf8("MED BOX") 
        AND #lineitem.l_quantity < (
            Subquery: Projection: Float64(0.2) * #AVG(lineitem.l_quantity)
                Aggregate: groupBy=[[]], aggr=[[AVG(#lineitem.l_quantity)]]
                    Filter: #lineitem.l_partkey = #part.p_partkey
                        TableScan: lineitem
        )
      Inner Join: #lineitem.l_partkey = #part.p_partkey
        TableScan: lineitem
        TableScan: part
```

because the indents aren't trustworthy when subqueries are taken into account.

Also, I've had a hard time writing unit tests for optimizers because building them programmatically is verbose and error-prone.

**Describe the solution you'd like**

If we were to create a visitor method like `display_json()`, and an equivalent `LogicalPlan::from_json()` then both problems would be solved fairly elegantly. As a bonus, plans would be readable and manipulable by common off the the shelf tools and readable by the uninitiated. 

**Describe alternatives you've considered**

Fixing the indentation on the formatter, writing different types of tests, crying into my beer.

### PR DESCRIPTION
**Pull request by tustvold**
# Which issue does this PR close?

Closes #2889

 # Rationale for this change

This provides a way to serialize the datafusion-proto messages to any serde serializer, e.g. JSON, based on the protobuf [JSON specification](https://developers.google.com/protocol-buffers/docs/proto3#json). 

# What changes are included in this PR?

Uses pbjson to auto-generate the serde implementations for the protobuf messages. Full disclosure: I am the primary author and maintainer of this crate.

# Are there any user-facing changes?

No

### CODE REVIEW

File: datafusion/proto/src/lib.rs [hunk 1]
```diff
@@ -75,19 +78,32 @@ mod roundtrip_tests {
...
+
+    #[cfg(not(feature = "serde"))]
+    fn roundtrip_serde_test(_proto: &protobuf::LogicalExprNode) {}
+
     // Given a DataFusion logical Expr, convert it to protobuf and back, using debug formatting to test
     // equality.
-    macro_rules! roundtrip_expr_test {
-        ($initial_struct:ident, $ctx:ident) => {
-            let proto: protobuf::LogicalExprNode = (&$initial_struct).try_into().unwrap();
+    fn roundtrip_expr_test<T, E>(initial_struct: T, ctx: SessionContext)
```
Comment (tustvold) [inline datafusion/proto/src/lib.rs:93]: Drive by refactor to eliminate an unnecessary macro

File: datafusion/proto/src/lib.rs [hunk 2]
```diff
@@ -75,19 +78,32 @@ mod roundtrip_tests {
...
+    where
+        for<'a> &'a T: TryInto<protobuf::LogicalExprNode, Error = E> + Debug,
+        E: Debug,
+    {
+        let proto: protobuf::LogicalExprNode = (&initial_struct).try_into().unwrap();
+        let round_trip: Expr = parse_expr(&proto, &ctx).unwrap();
```
[the remark this answers is not in the document]
Comment (avantgardnerio) [reply datafusion/proto/src/lib.rs:99]: Oh, I was running my tests without the "serde" feature :facepalm:

[the remark this answers is not in the document]
Comment (avantgardnerio) [reply datafusion/proto/src/lib.rs:99]: Okay, I got all that sorted, and was able to do what I was trying to use this PR for: https://github.com/tustvold/arrow-datafusion/pull/63

I was hoping (naively) to see something like:

```
{Projection: { field: "#employee_csv.id",
  Filter: {expr: {left: "#employee_csv.state", op: "IN", right: {
    Subquery: {
      TableScan: {name: "employee_csv", projection=["state"]},
    TableScan: {name: "employee_csv" projection=["id", "state"]
}}
```

Unfortunately (though serializable and deserializable) the result is not as human readable as I hoped, and probably not something I'd want to check into a test assertion: https://gist.github.com/avantgardnerio/35a04950ca768fdfe6579aea08126b74

[the remark this answers is not in the document]
Comment (avantgardnerio) [reply datafusion/proto/src/lib.rs:99]: Running that raw JSON through [this formatter](https://github.com/lydell/json-stringify-pretty-compact) produces some encouraging output:

```
{
  "projection": {
    "input": {
      "sort": {
        "input": {
          "projection": {
            "input": {
              "listingScan": {
                "tableName": "alltypes_plain",
                "paths": ["file:///home/bgardner/workspace/arrow-datafusion/parquet-testing/data/alltypes_plain.parquet"],
                "fileExtension": ".parquet",
                "schema": {
                  "columns": [
                    {"name": "id", "arrowType": {"INT32": {}}, "nullable": true},
                    {"name": "bool_col", "arrowType": {"BOOL": {}}, "nullable": true},
                    {"name": "tinyint_col", "arrowType": {"INT32": {}}, "nullable": true},
                    {"name": "smallint_col", "arrowType": {"INT32": {}}, "nullable": true},
                    {"name": "int_col", "arrowType": {"INT32": {}}, "nullable": true},
                    {"name": "bigint_col", "arrowType": {"INT64": {}}, "nullable": true},
                    {"name": "float_col", "arrowType": {"FLOAT32": {}}, "nullable": true},
                    {"name": "double_col", "arrowType": {"FLOAT64": {}}, "nullable": true},
                    {"name": "date_string_col", "arrowType": {"BINARY": {}}, "nullable": true},
                    {"name": "string_col", "arrowType": {"BINARY": {}}, "nullable": true},
                    {"name": "timestamp_col", "arrowType": {"TIMESTAMP": {"timeUnit": "Nanosecond"}}, "nullable": true}
                  ]
                },
                "collectStat": true,
                "targetPartitions": 16,
                "parquet": {"enablePruning": true}
              }
            },
            "expr": [
              {"column": {"name": "id", "relation": {"relation": "alltypes_plain"}}},
              {"column": {"name": "int_col", "relation": {"relation": "alltypes_plain"}}},
              {"column": {"name": "double_col", "relation": {"relation": "alltypes_plain"}}}
            ]
          }
        },
        "expr": [
          {"sort": {"expr": {"column": {"name": "int_col", "relation": {"relation": "alltypes_plain"}}}, "asc": true}},
          {"sort": {"expr": {"column": {"name": "double_col", "relation": {"relation": "alltypes_plain"}}}, "asc": true}}
        ]
      }
    },
    "expr": [{"column": {"name": "id", "relation": {"relation": "alltypes_plain"}}}]
  }
}
```

**Review by avantgardnerio — APPROVED** [review]
After messing around a bit, I'm convinced this is just what we need. It does look like the `DefaultExtensionCodec` is not exposed in other packages? So maybe it makes sense to expose methods like `logical_plan_to_json()` for similar reasons as `logical_plan_to_bytes()` - making it easier to use from outside the `proto` crate...

**Review by avantgardnerio — CHANGES_REQUESTED** [review]
Please take a look at https://github.com/tustvold/arrow-datafusion/pull/64 before merging into Apache. I think those couple tweaks are important.

**Review by avantgardnerio — APPROVED** [review]

File: datafusion/proto/Cargo.toml [hunk 3]
```diff
@@ -33,18 +33,24 @@ name = "datafusion_proto"
...
+serde_json = { version = "1.0", optional = true }
+pbjson = { version = "0.3", optional = true }
+pbjson-types = { version = "0.3", optional = true }
 
 [dev-dependencies]
 doc-comment = "0.3"
 tokio = "1.18"
 
 [build-dependencies]
 tonic-build = { version = "0.7" }
```
Comment (avantgardnerio) [inline datafusion/proto/Cargo.toml:55]:
```suggestion
pbjson-build = { version = "0.3", optional = true }
```

File: datafusion/proto/Cargo.toml [hunk 4]
```diff
@@ -33,18 +33,24 @@ name = "datafusion_proto"
...
+pbjson-build = { version = "0.3", optional = true }
```
Comment (avantgardnerio) [inline datafusion/proto/Cargo.toml:56]:
```suggestion
tonic-build = { version = "0.7" }
```

File: datafusion/proto/Cargo.toml [hunk 5]
```diff
@@ -33,18 +33,24 @@ name = "datafusion_proto"
...
+pbjson-types = { version = "0.3", optional = true }
 
 [dev-dependencies]
 doc-comment = "0.3"
 tokio = "1.18"
 
 [build-dependencies]
-tonic-build = { version = "0.7" }
+pbjson-build = { version = "0.3", optional = true }
+prost-build = { version = "0.7" }
```
Comment (tustvold) [inline datafusion/proto/Cargo.toml:56]: There was actually no reason for this to depend on tonic-build, and not just prost-build. This is likely an orphan from when ballista was extracted

File: datafusion/proto/Cargo.toml [hunk 6]
```diff
@@ -33,18 +33,24 @@ name = "datafusion_proto"
...
 
 [features]
+default = []
+json = ["pbjson", "pbjson-build", "serde", "serde_json"]
 
 [dependencies]
 arrow = { version = "18.0.0" }
 datafusion = { path = "../core", version = "10.0.0" }
 datafusion-common = { path = "../common", version = "10.0.0" }
 datafusion-expr = { path = "../expr", version = "10.0.0" }
```
Comment (avantgardnerio) [inline datafusion/proto/Cargo.toml:43]:
```suggestion
pbjson = { version = "0.3", optional = true }
pbjson-types = { version = "0.3", optional = true }
```

Comment (avantgardnerio) [inline datafusion/proto/Cargo.toml:43]:
```suggestion
datafusion-expr = { path = "../expr", version = "10.0.0" }
pbjson = { version = "0.3", optional = true }
pbjson-types = { version = "0.3", optional = true }
```

File: datafusion/proto/Cargo.toml [hunk 7]
```diff
@@ -33,18 +33,24 @@ name = "datafusion_proto"
...
 prost = "0.10"
-
+serde = { version = "1.0", optional = true }
+serde_json = { version = "1.0", optional = true }
+pbjson = { version = "0.3", optional = true }
+pbjson-types = { version = "0.3", optional = true }
```
Comment (avantgardnerio) [inline datafusion/proto/Cargo.toml:48]:
```suggestion
```

File: datafusion/proto/Cargo.toml [hunk 8]
```diff
@@ -33,18 +33,24 @@ name = "datafusion_proto"
...
 [dependencies]
 arrow = { version = "18.0.0" }
 datafusion = { path = "../core", version = "10.0.0" }
 datafusion-common = { path = "../common", version = "10.0.0" }
 datafusion-expr = { path = "../expr", version = "10.0.0" }
 prost = "0.10"
-
+serde = { version = "1.0", optional = true }
+serde_json = { version = "1.0", optional = true }
+pbjson = { version = "0.3", optional = true }
```
Comment (avantgardnerio) [inline datafusion/proto/Cargo.toml:47]:
```suggestion
```

**Review by avantgardnerio — COMMENTED** [review]
Cargo fix sugestions

**Pull request merged**